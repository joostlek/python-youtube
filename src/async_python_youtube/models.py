"""Models for YouTube API."""
from datetime import datetime
from typing import TypeVar

from pydantic import BaseModel, Field

from async_python_youtube.const import LiveBroadcastContent

T = TypeVar("T")


class YouTubeVideoThumbnail(BaseModel):
    """Model representing a video thumbnail."""

    url: str = Field(...)
    width: int = Field(...)
    height: int = Field(...)


class YouTubeVideoThumbnails(BaseModel):
    """Model representing video thumbnails."""

    default: YouTubeVideoThumbnail = Field(...)
    medium: YouTubeVideoThumbnail = Field(...)
    high: YouTubeVideoThumbnail = Field(...)
    standard: YouTubeVideoThumbnail = Field(...)
    maxres: YouTubeVideoThumbnail | None = Field(None)


class YouTubeVideoSnippet(BaseModel):
    """Model representing video snippet."""

    published_at: datetime = Field(..., alias="publishedAt")
    channel_id: str = Field(..., alias="channelId")
    title: str = Field(...)
    description: str = Field(...)
    thumbnails: YouTubeVideoThumbnails = Field(...)
    channel_title: str = Field(..., alias="channelTitle")
    tags: list[str] = Field(...)
    live_broadcast_content: LiveBroadcastContent = Field(
        ...,
        alias="liveBroadcastContent",
    )
    default_language: str | None = Field(None, alias="defaultLanguage")
    default_audio_language: str | None = Field(None, alias="defaultAudioLanguage")


class YouTubeVideo(BaseModel):
    """Model representing a video."""

    video_id: str = Field(..., alias="id")
    snippet: YouTubeVideoSnippet | None = None
