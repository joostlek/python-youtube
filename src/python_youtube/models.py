"""The YouTube API."""
from datetime import datetime

from pydantic import BaseModel, Field


class YouTubeVideoThumbnail(BaseModel):
    """Model representing a video thumbnail."""

    url: str = Field(...)
    width: int = Field(...)
    height: int = Field(...)


class YouTubeVideoThumbnails(BaseModel):
    """Model representing video thumbnails."""

    default: YouTubeVideoThumbnail = Field(...)
    medium: YouTubeVideoThumbnail = Field(...)
    high: YouTubeVideoThumbnail = Field(...)
    standard: YouTubeVideoThumbnail = Field(...)
    maxres: YouTubeVideoThumbnail = Field(...)


class YouTubeVideoSnippet(BaseModel):
    """Model representing video snippet."""

    published_at: datetime = Field(...)
    channel_id: str = Field(...)
    title: str = Field(...)
    description: str = Field(...)
    thumbnails: dict[str, YouTubeVideoThumbnail] = Field(...)


class YouTubeVideo(BaseModel):
    """Model representing a video."""

    snippet: YouTubeVideoSnippet | None = None
