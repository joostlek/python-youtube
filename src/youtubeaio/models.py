"""Models for YouTube API."""
from datetime import datetime
from typing import TypeVar

from pydantic import BaseModel, Field

from youtubeaio.const import LiveBroadcastContent

__all__ = [
    "YouTubeThumbnail",
    "YouTubeVideoThumbnails",
    "YouTubeVideoSnippet",
    "YouTubeVideo",
    "YouTubeChannelThumbnails",
    "YouTubeChannelRelatedPlaylists",
    "YouTubeChannelContentDetails",
    "YouTubeChannelSnippet",
    "YouTubeChannel",
]

T = TypeVar("T")


class YouTubeThumbnail(BaseModel):
    """Model representing a video thumbnail."""

    url: str = Field(...)
    width: int = Field(...)
    height: int = Field(...)


class YouTubeVideoThumbnails(BaseModel):
    """Model representing video thumbnails."""

    default: YouTubeThumbnail = Field(...)
    medium: YouTubeThumbnail = Field(...)
    high: YouTubeThumbnail = Field(...)
    standard: YouTubeThumbnail = Field(...)
    maxres: YouTubeThumbnail | None = Field(None)


class YouTubeVideoSnippet(BaseModel):
    """Model representing video snippet."""

    published_at: datetime = Field(..., alias="publishedAt")
    channel_id: str = Field(..., alias="channelId")
    title: str = Field(...)
    description: str = Field(...)
    thumbnails: YouTubeVideoThumbnails = Field(...)
    channel_title: str = Field(..., alias="channelTitle")
    tags: list[str] = Field([])
    live_broadcast_content: LiveBroadcastContent = Field(
        ...,
        alias="liveBroadcastContent",
    )
    default_language: str | None = Field(None, alias="defaultLanguage")
    default_audio_language: str | None = Field(None, alias="defaultAudioLanguage")


class YouTubeVideo(BaseModel):
    """Model representing a video."""

    video_id: str = Field(..., alias="id")
    snippet: YouTubeVideoSnippet | None = None


class YouTubeChannelThumbnails(BaseModel):
    """Model representing channel thumbnails."""

    default: YouTubeThumbnail = Field(...)
    medium: YouTubeThumbnail = Field(...)
    high: YouTubeThumbnail = Field(...)


class YouTubeChannelRelatedPlaylists(BaseModel):
    """Model representing related playlists of a channel."""

    likes: str = Field(...)
    uploads: str = Field(...)


class YouTubeChannelContentDetails(BaseModel):
    """Model representing content details of a channel."""

    related_playlists: YouTubeChannelRelatedPlaylists = Field(
        ...,
        alias="relatedPlaylists",
    )


class YouTubeChannelSnippet(BaseModel):
    """Model representing channel snippet."""

    title: str = Field(...)
    description: str = Field(...)
    published_at: datetime = Field(..., alias="publishedAt")
    thumbnails: YouTubeChannelThumbnails = Field(...)
    default_language: str | None = Field(None, alias="defaultLanguage")


class YouTubeChannel(BaseModel):
    """Model representing a YouTube channel."""

    channel_id: str = Field(..., alias="id")
    snippet: YouTubeChannelSnippet | None = None
    content_details: YouTubeChannelContentDetails | None = Field(
        None,
        alias="contentDetails",
    )
