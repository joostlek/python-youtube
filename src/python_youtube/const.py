"""Models for YouTube API."""
from enum import Enum


class HttpStatusCode(int, Enum):
    """Enum holding http status codes."""

    NOT_FOUND = 404


class VideoPart(str, Enum):
    """Enum holding the part parameters for video requests."""

    CONTENT_DETAILS = "contentDetails"
    FILE_DETAILS = "fileDetails"
    ID = "id"
    LIVE_STREAMING_DETAILS = "liveStreamingDetails"
    LOCALIZATIONS = "localizations"
    PLAYER = "player"
    PROCESSING_DETAILS = "processingDetails"
    RECORDING_DETAILS = "recordingDetails"
    SNIPPET = "snippet"
    STATISTICS = "statistics"
    STATUS = "status"
    SUGGESTIONS = "suggestions"
    TOPIC_DETAILS = "topicDetails"


class LiveBroadcastContent(str, Enum):
    """Enum holding the liveBroadcastContent values."""

    NONE = "none"
    LIVE = "live"
    UPCOMING = "upcoming"
