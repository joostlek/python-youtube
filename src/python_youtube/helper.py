"""Helper functions for the YouTube API."""
from collections.abc import AsyncGenerator, Generator
from typing import TypeVar

T = TypeVar("T")


async def first(generator: AsyncGenerator[T, None]) -> T | None:
    """Return the first value or None from the given AsyncGenerator."""
    try:
        return await generator.__anext__()
    except StopAsyncIteration:
        return None


def chunk(source: list[T], chunk_size: int) -> Generator[list[T], None, None]:
    """Divide the source list in chunks of given size."""
    for i in range(0, len(source), chunk_size):
        yield source[i : i + chunk_size]


async def limit(
    generator: AsyncGenerator[T, None],
    total: int,
) -> AsyncGenerator[T, None]:
    """Limit the number of entries returned from the AsyncGenerator."""
    if total < 1:
        msg = "Limit has to be an int > 1"
        raise ValueError(msg)
    count = 0
    async for item in generator:
        count += 1
        if count > total:
            break
        yield item
