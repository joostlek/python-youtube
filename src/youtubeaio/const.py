"""Models for YouTube API."""
from enum import Enum


class HttpStatusCode(int, Enum):
    """Enum holding http status codes."""

    NOT_FOUND = 404


class VideoPart(str, Enum):
    """Enum holding the part parameters for video requests."""

    CONTENT_DETAILS = "contentDetails"
    FILE_DETAILS = "fileDetails"
    ID = "id"
    LIVE_STREAMING_DETAILS = "liveStreamingDetails"
    LOCALIZATIONS = "localizations"
    PLAYER = "player"
    PROCESSING_DETAILS = "processingDetails"
    RECORDING_DETAILS = "recordingDetails"
    SNIPPET = "snippet"
    STATISTICS = "statistics"
    STATUS = "status"
    SUGGESTIONS = "suggestions"
    TOPIC_DETAILS = "topicDetails"


class VideoDimension(str, Enum):
    """Enum holding the possible video dimensions."""

    D3 = "3d"
    D2 = "2d"


class VideoDefinition(str, Enum):
    """Enum holding the possible video definitions."""

    HD = "hd"
    SD = "sd"


class VideoProjection(str, Enum):
    """Enum holding the possible video projections."""

    THREE_SIXTY = "360"
    RECTANGULAR = "rectangular"


class LiveBroadcastContent(str, Enum):
    """Enum holding the liveBroadcastContent values."""

    NONE = "none"
    LIVE = "live"
    UPCOMING = "upcoming"
