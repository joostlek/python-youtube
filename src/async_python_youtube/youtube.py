"""The YouTube API."""

import asyncio
from collections.abc import AsyncGenerator
from dataclasses import dataclass
from importlib import metadata
from typing import Any, cast

import async_timeout
from aiohttp import ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from async_python_youtube.const import HttpStatusCode
from async_python_youtube.exceptions import (
    YouTubeConnectionError,
    YouTubeError,
    YouTubeNotFoundError,
)
from async_python_youtube.helper import chunk, first
from async_python_youtube.models import YouTubeVideo

__all__ = [
    "YouTube",
]

MAX_RESULTS_FOR_VIDEO = 50


@dataclass
class YouTube:
    """YouTube API client."""

    session: ClientSession | None = None
    request_timeout: int = 10
    api_host: str = "youtube.googleapis.com"
    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        data: dict[str, Any] | None = None,
        error_handler: dict[int, BaseException] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to OpenSky.

        A generic method for sending/handling HTTP requests done against
        OpenSky.

        Args:
        ----
            uri: the path to call.
            data: the query parameters to add.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the API.

        Raises:
        ------
            OpenSkyConnectionError: An error occurred while communicating with
                the OpenSky API.
            OpenSkyrror: Received an unexpected response from the OpenSky API.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host=self.api_host,
            port=443,
            path="/youtube/v3/",
        ).joinpath(uri)

        headers = {
            "User-Agent": f"PythonOpenSky/{version}",
            "Accept": "application/json, text/plain, */*",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    METH_GET,
                    url.with_query(data),
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the YouTube API"
            raise YouTubeConnectionError(msg) from exception
        except ClientResponseError as exception:
            if error_handler and exception.status in error_handler:
                raise error_handler[exception.status] from exception
            msg = "Error occurred while communicating with YouTube API"
            raise YouTubeConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from the YouTube API"
            raise YouTubeError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def get_video(self, video_id: str) -> YouTubeVideo | None:
        """Get a single video."""
        return await first(self.get_videos([video_id]))

    async def get_videos(
        self,
        video_ids: list[str],
    ) -> AsyncGenerator[YouTubeVideo, None]:
        """Get a list of videos."""
        error_handler: dict[int, BaseException] = {
            HttpStatusCode.NOT_FOUND: YouTubeNotFoundError("Video not found"),
        }
        for video_chunk in chunk(video_ids, MAX_RESULTS_FOR_VIDEO):
            ids = ",".join(video_chunk)
            data = {
                "part": "snippet",
                "id": ids,
                "maxResults": MAX_RESULTS_FOR_VIDEO,
            }
            res = await self._request("videos", data=data, error_handler=error_handler)
            for item in res["items"]:
                yield YouTubeVideo.parse_obj(item)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Any:
        """Async enter.

        Returns
        -------
            The YouTube object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
