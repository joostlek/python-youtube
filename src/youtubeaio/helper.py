"""Helper functions for the YouTube API."""
import urllib.parse
from collections.abc import AsyncGenerator, Generator
from enum import Enum
from typing import Any, TypeVar

from youtubeaio.types import AuthScope

__all__ = [
    "YOUTUBE_AUTH_BASE_URL",
    "YOUTUBE_AUTH_TOKEN_URL",
    "build_scope",
    "build_url",
    "first",
    "chunk",
    "limit",
]

T = TypeVar("T")

YOUTUBE_AUTH_BASE_URL: str = "https://oauth2.googleapis.com"
YOUTUBE_AUTH_TOKEN_URL: str = f"{YOUTUBE_AUTH_BASE_URL}/token"


def build_scope(scopes: list[AuthScope]) -> str:
    """Build a valid scope string from list.

    :param scopes: list of :class:`~youtubeaio.types.AuthScope`
    :returns: the valid auth scope string
    """
    return " ".join([s.value for s in scopes])


def build_url(
    url: str,
    params: dict[str, Any],
    remove_none: bool = False,
    split_lists: bool = False,
    enum_value: bool = True,
) -> str:
    """Build a valid url string.

    :param url: base URL
    :param params: dictionary of URL parameter
    :param remove_none: if set all params that have a None value get removed
    |default| :code:`False`
    :param split_lists: if set all params that are a list will be split over multiple
    url parameter with the same name |default| :code:`False`
    :param enum_value: if true, automatically get value string from Enum values
    |default| :code:`True`
    :return: URL
    """

    def get_value(input_value: Any) -> str:
        if not enum_value:
            return str(input_value)
        if isinstance(input_value, Enum):
            return str(input_value.value)
        return str(input_value)

    def add_param(res: str, query_key: str, query_value: Any) -> str:
        if len(res) > 0:
            res += "&"
        res += query_key
        if query_value is not None:
            res += "=" + urllib.parse.quote(get_value(query_value))
        return res

    result = ""
    for key, value in params.items():
        if value is None and remove_none:
            continue
        if split_lists and isinstance(value, list):
            for val in value:
                result = add_param(result, key, val)
        else:
            result = add_param(result, key, value)
    return url + (("?" + result) if len(result) > 0 else "")


async def first(generator: AsyncGenerator[T, None]) -> T | None:
    """Return the first value or None from the given AsyncGenerator."""
    try:
        return await generator.__anext__()
    except StopAsyncIteration:
        return None


def chunk(source: list[T], chunk_size: int) -> Generator[list[T], None, None]:
    """Divide the source list in chunks of given size."""
    for i in range(0, len(source), chunk_size):
        yield source[i : i + chunk_size]


async def limit(
    generator: AsyncGenerator[T, None],
    total: int,
) -> AsyncGenerator[T, None]:
    """Limit the number of entries returned from the AsyncGenerator."""
    if total < 1:
        msg = "Limit has to be an int > 1"
        raise ValueError(msg)
    count = 0
    async for item in generator:
        count += 1
        if count > total:
            break
        yield item
