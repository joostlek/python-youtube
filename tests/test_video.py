"""Tests for the YouTube client."""
import json
from datetime import timedelta

import aiohttp
import pytest
from aiohttp.web_request import BaseRequest
from aresponses import Response, ResponsesMockServer
from syrupy import SnapshotAssertion

from youtubeaio.helper import first
from youtubeaio.models import YouTubeVideoThumbnails
from youtubeaio.types import PartMissingError
from youtubeaio.youtube import YouTube

from . import construct_fixture, load_fixture
from .const import YOUTUBE_URL
from .helper import get_thumbnail


async def test_fetch_video(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving a video."""
    aresponses.add(
        YOUTUBE_URL,
        "/youtube/v3/videos",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=json.dumps(
                construct_fixture("video", ["snippet", "contentDetails"], 1),
            ),
        ),
    )
    async with aiohttp.ClientSession() as session, YouTube(session=session) as youtube:
        video = await youtube.get_video(video_id="Ks-_Mh1QhMc")
        assert video == snapshot


async def test_fetch_videos(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving a list of videos."""

    async def response_handler(req: BaseRequest) -> Response:
        """Response handler for this test."""
        if "pageToken" in req.query:
            fixture = "video_response_2.json"
        else:
            fixture = "video_response_1.json"
        return aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(fixture),
        )

    aresponses.add(
        YOUTUBE_URL,
        "/youtube/v3/videos",
        "GET",
        response_handler,
        repeat=2,
    )
    async with aiohttp.ClientSession() as session:
        youtube = YouTube(session=session)
        videos = youtube.get_videos(
            video_ids=["Ks-_Mh1QhMc", "GvgqDSnpRQM", "V4DDt30Aat4"],
        )
        video1 = await videos.__anext__()
        assert video1
        assert video1.video_id == "Ks-_Mh1QhMc"
        video2 = await videos.__anext__()
        assert video2
        assert video2.video_id == "GvgqDSnpRQM"
        video3 = await videos.__anext__()
        assert video3
        assert video3.video_id == "V4DDt30Aat4"


async def test_fetch_single_page_video(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving a page of videos."""
    aresponses.add(
        YOUTUBE_URL,
        "/youtube/v3/videos",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("video_response_2.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        youtube = YouTube(session=session)
        videos = youtube.get_videos(
            video_ids=["V4DDt30Aat4"],
        )
        video3 = await videos.__anext__()
        assert video3
        assert video3.video_id == "V4DDt30Aat4"
        with pytest.raises(StopAsyncIteration):
            await videos.__anext__()


async def test_fetch_no_videos() -> None:
    """Test retrieving no videos."""
    youtube = YouTube()
    with pytest.raises(ValueError):
        await first(youtube.get_videos(video_ids=[]))


@pytest.mark.parametrize(
    ("thumbnails", "result_url"),
    [
        (
            YouTubeVideoThumbnails(
                maxres=get_thumbnail("maxres"),
                standard=get_thumbnail("standard"),
                high=get_thumbnail("high"),
                medium=get_thumbnail("medium"),
                default=get_thumbnail("default"),
            ),
            "maxres",
        ),
        (
            YouTubeVideoThumbnails(
                maxres=None,
                standard=get_thumbnail("standard"),
                high=get_thumbnail("high"),
                medium=get_thumbnail("medium"),
                default=get_thumbnail("default"),
            ),
            "standard",
        ),
        (
            YouTubeVideoThumbnails(
                maxres=None,
                standard=None,
                high=get_thumbnail("high"),
                medium=get_thumbnail("medium"),
                default=get_thumbnail("default"),
            ),
            "high",
        ),
        (
            YouTubeVideoThumbnails(
                maxres=None,
                standard=None,
                high=None,
                medium=get_thumbnail("medium"),
                default=get_thumbnail("default"),
            ),
            "medium",
        ),
        (
            YouTubeVideoThumbnails(
                maxres=None,
                standard=None,
                high=None,
                medium=None,
                default=get_thumbnail("default"),
            ),
            "default",
        ),
    ],
)
async def test_get_hq_thumbnail(
    thumbnails: YouTubeVideoThumbnails,
    result_url: str,
) -> None:
    """Check if the highest quality thumbnail is returned."""
    assert thumbnails.get_highest_quality().url == result_url


async def test_nullable_fields(
    aresponses: ResponsesMockServer,
) -> None:
    """Check if the fields exist when they are filled."""
    aresponses.add(
        YOUTUBE_URL,
        "/youtube/v3/videos",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=json.dumps(
                construct_fixture("video", ["snippet", "contentDetails"], 1),
            ),
        ),
    )
    async with aiohttp.ClientSession() as session:
        youtube = YouTube(session=session)
        video = await youtube.get_video(video_id="V4DDt30Aat4")
        assert video
        assert video.snippet.channel_id == "UCAuUUnT6oDeKwE6v1NGQxug"
        assert video.content_details.duration == timedelta(minutes=21, seconds=3)
        assert video.content_details.caption is True


async def test_nullable_fields_null(
    aresponses: ResponsesMockServer,
) -> None:
    """Check if an error is thrown if a non-requested part is accessed."""
    aresponses.add(
        YOUTUBE_URL,
        "/youtube/v3/videos",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=json.dumps(construct_fixture("video", [], 1)),
        ),
    )
    async with aiohttp.ClientSession() as session:
        youtube = YouTube(session=session)
        video = await youtube.get_video(video_id="V4DDt30Aat4")
        assert video
        with pytest.raises(PartMissingError):
            assert video.snippet.thumbnails
        with pytest.raises(PartMissingError):
            assert video.content_details
