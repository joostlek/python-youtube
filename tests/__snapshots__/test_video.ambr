# serializer version: 1
# name: test_fetch_video
  YouTubeVideo(video_id='Ks-_Mh1QhMc', nullable_snippet=YouTubeVideoSnippet(published_at=datetime.datetime(2012, 10, 1, 15, 27, 35, tzinfo=TzInfo(UTC)), channel_id='UCAuUUnT6oDeKwE6v1NGQxug', title='Your body language may shape who you are | Amy Cuddy', description='Body language affects how others see us, but it may also change how we see ourselves. Social psychologist Amy Cuddy argues that "power posing" -- standing in a posture of confidence, even when we don\'t feel confident -- can boost feelings of confidence, and might have an impact on our chances for success. (Note: Some of the findings presented in this talk have been referenced in an ongoing debate among social scientists about robustness and reproducibility. Read Amy Cuddy\'s response here: http://ideas.ted.com/inside-the-debate-about-power-posing-a-q-a-with-amy-cuddy/)\n\nGet TED Talks recommended just for you! Learn more at https://www.ted.com/signup.\n\nThe TED Talks channel features the best talks and performances from the TED Conference, where the world\'s leading thinkers and doers give the talk of their lives in 18 minutes (or less). Look for talks on Technology, Entertainment and Design -- plus science, business, global issues, the arts and more.\n\nFollow TED on Twitter: http://www.twitter.com/TEDTalks\nLike TED on Facebook: https://www.facebook.com/TED\n\nSubscribe to our channel: https://www.youtube.com/TED', thumbnails=YouTubeVideoThumbnails(default=YouTubeThumbnail(url='https://i.ytimg.com/vi/Ks-_Mh1QhMc/default.jpg', width=120, height=90), medium=YouTubeThumbnail(url='https://i.ytimg.com/vi/Ks-_Mh1QhMc/mqdefault.jpg', width=320, height=180), high=YouTubeThumbnail(url='https://i.ytimg.com/vi/Ks-_Mh1QhMc/hqdefault.jpg', width=480, height=360), standard=YouTubeThumbnail(url='https://i.ytimg.com/vi/Ks-_Mh1QhMc/sddefault.jpg', width=640, height=480), maxres=YouTubeThumbnail(url='https://i.ytimg.com/vi/Ks-_Mh1QhMc/maxresdefault.jpg', width=1280, height=720)), channel_title='TED', tags=['Amy Cuddy', 'TED', 'TEDTalk', 'TEDTalks', 'TED Talk', 'TED Talks', 'TEDGlobal', 'brain', 'business', 'psychology', 'self', 'success'], live_broadcast_content=<LiveBroadcastContent.NONE: 'none'>, default_language='en', default_audio_language='en'), nullable_content_details=YouTubeVideoContentDetails(raw_duration='PT21M3S', dimension=<VideoDimension.D2: '2d'>, definition=<VideoDefinition.HD: 'hd'>, raw_caption='true', licensed_content=True, projection=<VideoProjection.RECTANGULAR: 'rectangular'>))
# ---
